vllm/_custom_ops.py

```diff
@@ -532,6 +532,52 @@ def cutlass_scaled_mm_azp(a: torch.Tensor,
     return out
 
 
+def cutlass_compress_entry(a: torch.Tensor) \
+    -> Tuple[torch.Tensor, torch.Tensor]:
+    assert (a.dtype is torch.int8 or a.dtype is torch.float8_e4m3fn or \
+            a.dtype is torch.bfloat16 or a.dtype is torch.float16)
+
+    # e.dtype: torch.uint8 so elemsPerElemE = 8b / 2b_per_nz = 4
+    elemsPerElemE = 4
+
+    m = a.shape[0]
+    k = a.shape[1]
+    a_compressed = torch.empty((m, k // 2), dtype=a.dtype, device=a.device)
+    e = torch.empty((m, k // 2 // elemsPerElemE),
+                    dtype=torch.uint8,
+                    device=a.device)
+
+    if not (torch.ops._C.cutlass_compress_entry(a_compressed, e, a)):
+        raise ValueError
+
+    return a_compressed, e
+
+
+def cutlass_scaled_sparse_mm(
+        a: torch.Tensor,  # row-major activations
+        b: torch.Tensor,  # row-major weight matrix
+        e: torch.Tensor,
+        scale_a: torch.Tensor,
+        scale_b: torch.Tensor,
+        out_dtype: torch.dtype,
+        bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+    # assert (b.shape[0] % 16 == 0 and b.shape[1] % 16 == 0)
+    assert (out_dtype is torch.bfloat16 or out_dtype is torch.float16)
+    assert bias is None or bias.shape[0] == a.shape[0] \
+        and bias.dtype == out_dtype
+
+    a_t = a.t()
```

Comment:
nit: I don't hate this, but it would be really nice if we could push these transposes into the C++ code (similar to Machete)

Reply:
Using something like LayoutA_Transpose = typename cutlass::layout::LayoutTranspose<LayoutA>::type; ?

Reply:
replicating slack response here for record keeping:
can be done like:
https://github.com/vllm-project/vllm/blob/4816d20aa43fdc4abf66c28f6690a1953d8adbe9/csrc/quantization/machete/machete_mm_kernel.cuh#L251-L255
you may also be interested in using:
https://github.com/vllm-project/vllm/blob/4816d20aa43fdc4abf66c28f6690a1953d8adbe9/csrc/cutlass_extensions/torch_utils.hpp#L51-L96
example usage:
https://github.com/vllm-project/vllm/blob/4816d20aa43fdc4abf66c28f6690a1953d8adbe9/csrc/quantization/machete/machete_mm_kernel.cuh#L197-L198